[LWG 11] P2572R1 std::format fill character allowances (#6144)
cplusplus/draft: source/utilities.tex

\end{ncbnf}

\pnum
Field widths are specified in defnadj{field width}{units};
the number of column positions required to display a sequence of
characters in a terminal.
The \defnadj{minimum}{field width}
is the number of field width units a replacement field minimally requires of
the formatted sequence of characters produced for a format argument.
The \defnadj{estimated}{field width} is the number of field width units
that are required for the formatted sequence of characters
produced for a format argument independent of
the effects of the \fmtgrammarterm{width} option.
The \defnadj{padding}{width} is the greater of \tcode{0} and
the difference of the minimum field width and the estimated field width.

\begin{note}
The POSIX \tcode{wcswidth} function is an example of a function that,
given a string, returns the number of column positions required by
a terminal to display the string.
\end{note}

\pnum
The \defnadj{fill}{character} is the character denoted by
the \fmtgrammarterm{fill} option or,

> if the fill option is absent, the space character.

Reviewer: ```suggestion
if the \fmtgrammarterm{fill} option is absent, the space character.
```

> For a format specification in a Unicode encoding,
> the fill character corresponds to a single UCS scalar value.

Reviewer: Are the following changes, which the paper for #6126 made for existing wording, not required here?
```suggestion
For a format specification in UTF-8, UTF-16, or UTF-32,
the fill character corresponds to a single Unicode scalar value.
```

Reviewer: Opened #6155 for this.

Author: Added Unicode fixes in a separate edit

string s4 = format("{:*^6}", 'x');      // value of \tcode{s4} is \tcode{"**x***"}
string s5 = format("{:6d}", c);         // value of \tcode{s5} is \tcode{"\ \ \ 120"}
string s6 = format("{:6}", true);       // value of \tcode{s6} is \tcode{"true\ \ "}

> string s7 = format("{:*<6.3}", "123456");   // value of s7 is "123***"

Reviewer: I think the comments should be aligned. Those from the added strings and the preexisting ones. Just like the paper.
![1677093953](https://user-images.githubusercontent.com/21071787/220737457-e16693dd-ec95-4de5-bf18-29b87046da97.png)

Author: ok, done

> string s7 = format("{:*<6.3}", "123456");   // value of s7 is "123***"
> string s8 = format("{:02}", 1234);          // value of s8 is "1234"
> string s9 = format("{:*<}", "12");          // value of s9 is "12"
> string sA = format("{:*<6}", "12345678");   // value of sA is "12345678"

Reviewer: ```suggestion
string s7 = format("{:*<6.3}", "123456");   // value of \tcode{s7} is \tcode{"123***"}
string s8 = format("{:02}", 1234);          // value of \tcode{s8} is \tcode{"1234"}
string s9 = format("{:*<}", "12");          // value of \tcode{s9} is \tcode{"12"}
string sA = format("{:*<6}", "12345678");   // value of \tcode{sA} is \tcode{"12345678"}
```

> \end{codeblock}

Reviewer: See https://github.com/cplusplus/draft/pull/6139#discussion_r1114664091. This is so that future additions are known to follow the sequence.
```suggestion
%% FIXME: string sB = ... string sB from P2572R1 ... // value of \tcode{sB} is ...
%% FIXME: string sC = ... string sC from P2572R1 ... // value of \tcode{sC} is ...
\end{codeblock}
```

Author: Unfortunately we can't do that (have a FIXME inside the `\codeblock`) because it would appear in the final paper, and I'm pretty sure we don't want that.

\end{example}
\begin{note}

> Unless a minimum field width is defined, the field width is determined by
> the size of the content and the alignment option has no effect.
> 
> The \fmtgrammarterm{fill}, \fmtgrammarterm{align}, and \tcode{0} options

Reviewer: ```suggestion
The \fmtgrammarterm{fill}, \fmtgrammarterm{align}, and \tcode{0} options
```

trailing zeros are not removed from the result.

\pnum

> The (\tcode{0}) option is valid for arithmetic types

Author: thanks!

trailing zeros are not removed from the result.

\pnum

> A (\tcode{0}) option is valid for arithmetic types

Reviewer: ```suggestion
The \tcode{0} option is valid for arithmetic types
```

other than \tcode{charT} and \tcode{bool} or
when an integer presentation type is specified.
For formatting arguments that have a value
other than an infinity or a NaN,
this option pads the formatted argument by
inserting the \tcode{0} character $n$ times
following the sign or base prefix indicators (if any)
where $n$ is \tcode{0} if the \fmtgrammarterm{align} option is present and
is the padding width otherwise.
\begin{example}
\begin{codeblock}
char c = 120;
string s1 = format("{:+06d}", c);       // value of \tcode{s1} is \tcode{"+00120"}
string s2 = format("{:#06x}", 0xa);     // value of \tcode{s2} is \tcode{"0x000a"}
string s3 = format("{:<06}", -42);      // value of \tcode{s3} is \tcode{"-42\ \ \ "} (\tcode{0} has no effect)

> string s4 = format("{:06}", inf);       // value of \tcode{s4} is "   inf" (\tcode{0} has no effect)

Reviewer: ```suggestion
string s4 = format("{:06}", inf);       // value of \tcode{s4} is \tcode{"   inf"} (\tcode{0} has no effect)
```

\end{note}

\pnum

> For a sequence of characters in a Unicode encoding,

Reviewer: Are the following changes, which the paper for #6126 made for existing wording, not required here?
```suggestion
For a sequence of characters in UTF-8, UTF-16, or UTF-32,
```

\item \ucode{20000} -- \ucode{2fffd}
\item \ucode{30000} -- \ucode{3fffd}
\end{itemize}
The field width of all other code points is 1.

\pnum

> For a sequence of characters in a non-Unicode encoding,

Reviewer: Are the following changes, which the paper for #6126 made for existing wording (https://github.com/cplusplus/draft/pull/6126/files#diff-afa9f66c3857a24dc4092685af2cbf9cbc44bbd35afa4b0eb26a14a324b0999dL15018), not required here?
```suggestion
For a sequence of characters in neither UTF-8, UTF-16, nor UTF-32,
```

> the field width is unspecified.

Reviewer: Should `\unspec` be used here?

Author: no, \unspec is for library code (e.g. in a synopsis)

\pnum
The \fmtgrammarterm{precision} option is valid
for floating-point and string types.
For floating-point types,
the value of this option specifies the precision
to be used for the floating-point presentation type.
For string types,
this option specifies the longest prefix of the formatted argument
to be included in the replacement field such that
the field width of the prefix is no greater than the value of this option.

\pnum
If \fmtgrammarterm{nonnegative-integer} is used in
a \fmtgrammarterm{precision} option,
the value of the decimal integer is used as the value of the option.

> whose estimated width is no greater than the precision.

Reviewer: This line is part of the deleted paragraph.
```suggestion
```

Author: Thanks, must not have gotten enough sleep before starting this one :)
